fdbclient/include/fdbclient/BulkLoading.h

```diff
@@ -0,0 +1,203 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FDBCLIENT_BULKLOADING_H
+#define FDBCLIENT_BULKLOADING_H
+#pragma once
+
+#include "fdbclient/FDBTypes.h"
+#include "fdbrpc/fdbrpc.h"
+
+enum class BulkLoadPhase : uint8_t {
+	Invalid = 0, // Set by users
+	Triggered = 1, // Update when DD trigger a data move for the task
+	Running = 2, // Update atomically with updating KeyServer dest servers in startMoveKey
+	Complete = 3, // Update atomically with updating KeyServer src servers in finishMoveKey
+};
+
+enum class BulkLoadType : uint8_t {
+	Invalid = 0,
+	SST = 1,
+};
+
+enum class BulkLoadTransportMethod : uint8_t {
+	Invalid = 0,
+	CP = 1,
+};
+
+enum class BulkLoadInjectMethod : uint8_t {
+	Invalid = 0,
+	File = 1,
+};
+
+struct BulkLoadState {
+	constexpr static FileIdentifier file_identifier = 1384499;
+
+	BulkLoadState() = default;
+
+	BulkLoadState(BulkLoadType loadType, std::string folder)
+	  : loadType(loadType), folder(folder), phase(BulkLoadPhase::Invalid), taskId(UID()) {}
+
+	BulkLoadState(KeyRange range, BulkLoadType loadType, std::string folder)
+	  : range(range), loadType(loadType), folder(folder), phase(BulkLoadPhase::Invalid), taskId(UID()) {}
+
+	bool isValid() const { return loadType != BulkLoadType::Invalid; }
+
+	std::string toString() const {
+		std::string res =
+		    "BulkLoadState: [Range]: " + Traceable<KeyRangeRef>::toString(range) +
+		    ", [Type]: " + std::to_string(static_cast<uint8_t>(loadType)) +
+		    ", [TransportMethod]: " + std::to_string(static_cast<uint8_t>(transportMethod)) +
+		    ", [InjectMethod]: " + std::to_string(static_cast<uint8_t>(injectMethod)) +
+		    ", [Phase]: " + std::to_string(static_cast<uint8_t>(phase)) + ", [Folder]: " + folder +
+		    ", [DataFiles]: " + describe(dataFiles) + ", [SubmitTime]: " + std::to_string(submitTime) +
+		    ", [TriggerTime]: " + std::to_string(triggerTime) + ", [StartTime]: " + std::to_string(startTime) +
+		    ", [CompleteTime]: " + std::to_string(completeTime) + ", [RestartCount]: " + std::to_string(restartCount);
+		if (bytesSampleFile.present()) {
+			res = res + ", [ByteSampleFile]: " + bytesSampleFile.get();
+		}
+		if (dataMoveId.present()) {
+			res = res + ", [DataMoveId]: " + dataMoveId.get().toString();
+		}
+		res = res + ", [TaskId]: " + taskId.toString();
+		return res;
+	}
+
+	bool setTaskId(UID id) {
+		if (taskId.isValid() && taskId != id) {
```

Comment:
taskid, type, transportmethod and a few others should be immutable? These methods should be private for the data members that shouldn't change after construction?

Reply:
(Maybe can't be private because of serialization?)